src/compiler/checker.ts

```diff
@@ -6814,6 +6904,22 @@ module ts {
             return undefinedType;
         }
 
+        function checkAwaitExpression(node: AwaitExpression): Type {
+            // Grammar checking
+            if (!(node.parserContextFlags & ParserContextFlags.Await)) {
+                var parameter = getContainingParameter(node);
+                if (parameter && parameter.parserContextFlags & ParserContextFlags.Await) {
+                    grammarErrorAfterFirstToken(node, Diagnostics._0_expression_is_not_allowed_in_an_initializer, "await");

```

Comment:
We try to recover `await expr` as an await expression, the same as how we parse `yield expr`. Since the grammar clears `[Await]` on Initializer, it is possible we have an invalid await expression in an initializer here. I may change the check for the parameter to use ParserContextFlags.AsyncParameter instead.

From: Jason Freeman [mailto:notifications@github.com]
Sent: Friday, February 13, 2015 6:06 PM
To: Microsoft/TypeScript
Cc: Ron Buckton
Subject: Re: [TypeScript] Async Functions (#1781)


In src/compiler/checker.ts<https://github.com/Microsoft/TypeScript/pull/1781#discussion_r24710812>:

> @@ -6814,6 +6904,22 @@ module ts {

>              return undefinedType;

>          }

>

> +        function checkAwaitExpression(node: AwaitExpression): Type {

> +            // Grammar checking

> +            if (!(node.parserContextFlags & ParserContextFlags.Await)) {

> +                var parameter = getContainingParameter(node);

> +                if (parameter && parameter.parserContextFlags & ParserContextFlags.Await) {

> +                    grammarErrorAfterFirstToken(node, Diagnostics._0_expression_is_not_allowed_in_an_initializer, "await");

Also, how can the node not be in an await context, but the parameter is in an await context?

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/1781/files#r24710812>.